Adding batch script to archive raindrop bookmark URLs

yakticus opened the pull request:
creating batch script that makes sure that raindrop bookmarks are included in the wayback machine (web archive)
    - archive-links.py is the main command-line script that does the archiving
    - raindrop_api.py contains general-purpose APIs for accessing raindrop (extracted from extract-raindrop-bookmarks.py)
    - extract-raindrop-bookmarks.py - refactored to use the above module
    - wayback_api.py - general purpose API for accessing the web archive

TODO:
    - seems to have problems with URLs that have query params (at least this example doesn't work: https://www.nytimes.com/2017/02/17/opinion/a-party-to-the-russian-connection.html?smid=tw-share)
    - make the requests async (should be faster)

Commit:
creating batch script that makes sure that raindrop bookmarks are included in the wayback machine (web archive)
- archive-links.py is the main command-line script that does the archiving
- raindrop_api.py contains general-purpose APIs for accessing raindrop (extracted from extract-raindrop-bookmarks.py)
- extract-raindrop-bookmarks.py - refactored to use the above module
- wayback_api.py - general purpose API for accessing the web archive

TODO:
- doesn't handle exceptions. Instead it just crashes whenever a request fails.  Should implement retries and then
- seems to have problems with URLs that have query params (at least this example doesn't work: https://www.nytimes.com/2017/02/17/opinion/a-party-to-the-russian-connection.html?smid=tw-share)
- make the requests async (should be faster)

Commit:
fixed a few bugs
- archive-links no longer tries endlessly to archive links that have a problem
- links with problem archiving are logged as such
- restored some of the options in the raindrop script that were broken by refactoring

Commit:
updated README

jpowerwa commented on extract-raindrop-bookmarks.py:
```diff
@@ -3,65 +3,14 @@
-
-RAINDROP_BOOKMARK_API = 'https://raindrop.io/api/raindrops/{collection_id}'
-PAGE_SIZE = 40  # max number of results per page 
+DEFAUlT_PAGE_SIZE = 40
```
Typo? I assume you were intending DEFAULT_PAGE_SIZE as the name, in all caps.

yakticus replied:
yep. thanks.

jpowerwa commented on archive-links.py:
```diff
@@ -0,0 +1,67 @@
+		log.setLevel(logging.DEBUG)
+	else:
+		log.setLevel(logging.INFO)
+
```
Having log at the module level is actually a Python logging best practice. In non-script modules, you'll often see it declared right after import logging, and set to the logger for `__name__`. That works nicely given the scoped names of Python modules; setting the log level for `foo` also sets it for `foo.*`, `foo.*.*`, etc. I know it seems messy to have it at the module level; it takes some getting used to.

yakticus replied:
Thanks for filling me in. Globals and module-scoped stuff is a no-no in environments I'm used to. Soon I will learn the ways of Python. :)

jpowerwa commented on raindrop_api.py:
```diff
@@ -0,0 +1,56 @@
+from typing import List, Tuple
+import requests
+import logging
+
```
So here, standard Python practice is to do:

```
import logging
log = logging.getLogger(__name__)
```

And then within the module, calls are simply made to `log.debug`, `log.warn`, etc.

jpowerwa commented on raindrop_api.py:
```diff
@@ -0,0 +1,56 @@
+    """
+    uri = RAINDROP_BOOKMARK_API.format(collection_id=collection_id)
+    query_params = {'page': page_number, 'perpage': page_size}
+    if log is not None:
```
This should go away once the `log` argument is removed, as the logger will be initialized when the module is loaded.

jpowerwa commented on raindrop_api.py:
```diff
@@ -0,0 +1,56 @@
+    response_data = requests.get(uri, params=query_params).json()
+    items = response_data.get('items', [])
+    total_item_count = response_data.get('count')
+    if log is not None:
```
As above.

jpowerwa reviewed (changes requested):
Looks great overall. A few requested changes:
1. Restore the log pattern in extract-raindrop-bookmarks and repeat in archive-links.
2. Do not pass logger as an argument; declare module-level logger named `log` in raindrop_api.py.
3. Please use 4-space indents instead of 8.
4. Minor typo: `DEFAUlT` -> `DEFAULT`

yakticus commented:
@jpowerwa Thanks for the review, and the pointers in the right direction on logging. I'll get this PR updated sometime tonight.

Commit:
Addressing PR feedback. specifically:

- Restore the log pattern in extract-raindrop-bookmarks and repeat in archive-links.
- Do not pass logger as an argument; declare module-level logger named log in raindrop_api.py.
- Please use 4-space indents instead of 8.
- Minor typo: DEFAUlT -> DEFAULT

yakticus commented:
@jpowerwa updated with feedback incorporated.

jpowerwa reviewed (commented):
Looks great! Go ahead and merge to master and then delete the branch.
